Index: openssl-1.1.0~pre5/Configurations/20-debian.conf
===================================================================
--- /dev/null
+++ openssl-1.1.0~pre5/Configurations/20-debian.conf
@@ -0,0 +1,29 @@
+my $debian_cflags = `dpkg-buildflags --get CFLAGS` . `dpkg-buildflags --get CPPFLAGS` . "-Wa,--noexecstack -Wall";
+$debian_cflags =~ s/\n/ /g;
+my $debian_ldflags = `dpkg-buildflags --get LDFLAGS`;
+$debian_ldflags =~ s/\n/ /g;
+
+%targets = (
+	"debian" => {
+		cflags => $debian_cflags,
+		lflags => add($debian_ldflags, "-pthread"),
+	},
+	"debian-alpha" => {
+		inherit_from => [ "linux-alpha-gcc", "debian" ],
+	},
+	"debian-alpha-ev4" => {
+		inherit_from => [ "debian-alpha" ],
+		cflags => add("-mcpu=ev4"),
+	},
+	"debian-alpha-ev5" => {
+		inherit_from => [ "debian-alpha" ],
+		cflags => add("-mcpu=ev5"),
+	},
+	"debian-arm64" => {
+		inherit_from => [ "linux-aarch64", "debian" ],
+	},
+	"debian-amd64" => {
+		inherit_from => [ "linux-x86_64", "debian" ],
+	},
+);
+
