From: Kurt Roeckx <kurt@roeckx.be>
Subject: Change default bit size and digest
Date: Fri, 01 Nov 2013 20:47:14 +0100

Index: openssl-1.0.1e/apps/openssl.cnf
===================================================================
--- openssl-1.0.1e.orig/apps/openssl.cnf	2013-11-01 20:48:24.000000000 +0100
+++ openssl-1.0.1e/apps/openssl.cnf	2013-11-01 20:49:09.000000000 +0100
@@ -72,7 +72,7 @@
 
 default_days	= 365			# how long to certify for
 default_crl_days= 30			# how long before next CRL
-default_md	= default		# use public key default MD
+default_md	= sha256
 preserve	= no			# keep passed DN ordering
 
 # A few difference way of specifying how similar the request should look
@@ -103,7 +103,8 @@
 
 ####################################################################
 [ req ]
-default_bits		= 1024
+default_bits		= 2048
+default_md		= sha256
 default_keyfile 	= privkey.pem
 distinguished_name	= req_distinguished_name
 attributes		= req_attributes
