Index: openssl-1.0.0c/ssl/ssl.h
===================================================================
--- openssl-1.0.0c.orig/ssl/ssl.h	2011-04-02 10:57:15.000000000 +0000
+++ openssl-1.0.0c/ssl/ssl.h	2011-04-02 11:03:51.000000000 +0000
@@ -1187,7 +1187,9 @@
 }
 #endif
 
+#ifndef OPENSSL_NO_SSL2
 #include <openssl/ssl2.h>
+#endif
 #include <openssl/ssl3.h>
 #include <openssl/tls1.h> /* This is mostly sslv3 with a few tweaks */
 #include <openssl/dtls1.h> /* Datagram TLS */
