Index: openssl-1.0.1/crypto/Makefile
===================================================================
--- openssl-1.0.1.orig/crypto/Makefile	2011-12-10 01:37:55.000000000 +0000
+++ openssl-1.0.1/crypto/Makefile	2012-03-17 09:29:15.000000000 +0000
@@ -60,7 +60,7 @@
 	echo "  #define DATE \"`LC_ALL=C LC_TIME=C date`\""; \
 	echo '#endif' ) >buildinf.h
 
-x86cpuid.s:	x86cpuid.pl perlasm/x86asm.pl
+x86cpuid.S:	x86cpuid.pl perlasm/x86asm.pl
 	$(PERL) x86cpuid.pl $(PERLASM_SCHEME) $(CFLAGS) $(PROCESSOR) > $@
 
 applink.o:	$(TOP)/ms/applink.c
@@ -72,7 +72,7 @@
 uplink-x86.s:	$(TOP)/ms/uplink-x86.pl
 	$(PERL) $(TOP)/ms/uplink-x86.pl $(PERLASM_SCHEME) > $@
 
-x86_64cpuid.s: x86_64cpuid.pl;	$(PERL) x86_64cpuid.pl $(PERLASM_SCHEME) > $@
+x86_64cpuid.S: x86_64cpuid.pl;	$(PERL) x86_64cpuid.pl $(PERLASM_SCHEME) > $@
 ia64cpuid.s: ia64cpuid.S;	$(CC) $(CFLAGS) -E ia64cpuid.S > $@
 ppccpuid.s:	ppccpuid.pl;	$(PERL) ppccpuid.pl $(PERLASM_SCHEME) $@
 pariscid.s:	pariscid.pl;	$(PERL) pariscid.pl $(PERLASM_SCHEME) $@
Index: openssl-1.0.1/crypto/x86_64cpuid.pl
===================================================================
--- openssl-1.0.1.orig/crypto/x86_64cpuid.pl	2011-11-14 21:01:16.000000000 +0000
+++ openssl-1.0.1/crypto/x86_64cpuid.pl	2012-03-17 09:20:04.000000000 +0000
@@ -20,7 +20,11 @@
 .extern		OPENSSL_cpuid_setup
 .hidden		OPENSSL_cpuid_setup
 .section	.init
+#ifdef OPENSSL_PIC
+	call	OPENSSL_cpuid_setup\@PLT
+#else
 	call	OPENSSL_cpuid_setup
+#endif
 
 .hidden	OPENSSL_ia32cap_P
 .comm	OPENSSL_ia32cap_P,8,4
Index: openssl-1.0.1/crypto/des/asm/desboth.pl
===================================================================
--- openssl-1.0.1.orig/crypto/des/asm/desboth.pl	2001-10-24 21:20:56.000000000 +0000
+++ openssl-1.0.1/crypto/des/asm/desboth.pl	2012-03-17 09:20:04.000000000 +0000
@@ -16,6 +16,11 @@
 
 	&push("edi");
 
+	&call   (&label("pic_point0"));
+	&set_label("pic_point0");
+	&blindpop("ebp");
+	&add    ("ebp", "\$_GLOBAL_OFFSET_TABLE_+[.-" . &label("pic_point0") . "]");
+
 	&comment("");
 	&comment("Load the data words");
 	&mov($L,&DWP(0,"ebx","",0));
@@ -47,15 +52,21 @@
 	&mov(&swtmp(2),	(DWC(($enc)?"1":"0")));
 	&mov(&swtmp(1),	"eax");
 	&mov(&swtmp(0),	"ebx");
-	&call("DES_encrypt2");
+	&exch("ebx", "ebp");
+	&call("DES_encrypt2\@PLT");
+	&exch("ebx", "ebp");
 	&mov(&swtmp(2),	(DWC(($enc)?"0":"1")));
 	&mov(&swtmp(1),	"edi");
 	&mov(&swtmp(0),	"ebx");
-	&call("DES_encrypt2");
+	&exch("ebx", "ebp");
+	&call("DES_encrypt2\@PLT");
+	&exch("ebx", "ebp");
 	&mov(&swtmp(2),	(DWC(($enc)?"1":"0")));
 	&mov(&swtmp(1),	"esi");
 	&mov(&swtmp(0),	"ebx");
-	&call("DES_encrypt2");
+	&exch("ebx", "ebp");
+	&call("DES_encrypt2\@PLT");
+	&exch("ebx", "ebp");
 
 	&stack_pop(3);
 	&mov($L,&DWP(0,"ebx","",0));
Index: openssl-1.0.1/crypto/rc4/Makefile
===================================================================
--- openssl-1.0.1.orig/crypto/rc4/Makefile	2011-11-14 20:42:21.000000000 +0000
+++ openssl-1.0.1/crypto/rc4/Makefile	2012-03-17 09:20:04.000000000 +0000
@@ -44,7 +44,7 @@
 rc4-586.s:	asm/rc4-586.pl ../perlasm/x86asm.pl
 	$(PERL) asm/rc4-586.pl $(PERLASM_SCHEME) $(CFLAGS) > $@
 
-rc4-x86_64.s: asm/rc4-x86_64.pl
+rc4-x86_64.S: asm/rc4-x86_64.pl
 	$(PERL) asm/rc4-x86_64.pl $(PERLASM_SCHEME) > $@
 rc4-md5-x86_64.s:	asm/rc4-md5-x86_64.pl
 	$(PERL) asm/rc4-md5-x86_64.pl $(PERLASM_SCHEME) > $@
Index: openssl-1.0.1/crypto/rc4/asm/rc4-x86_64.pl
===================================================================
--- openssl-1.0.1.orig/crypto/rc4/asm/rc4-x86_64.pl	2011-07-01 14:13:52.000000000 +0000
+++ openssl-1.0.1/crypto/rc4/asm/rc4-x86_64.pl	2012-03-17 09:28:11.000000000 +0000
@@ -157,7 +157,11 @@
 	mov	-4($dat),$YY#b
 	cmpl	\$-1,256($dat)
 	je	.LRC4_CHAR
+#ifdef OPENSSL_PIC
+	mov	OPENSSL_ia32cap_P\@GOTPCREL(%rip),%r8d
+#else
 	mov	OPENSSL_ia32cap_P(%rip),%r8d
+#endif
 	xor	$TX[1],$TX[1]
 	inc	$XX[0]#b
 	sub	$XX[0],$TX[1]
@@ -442,7 +446,11 @@
 	xor	%r10,%r10
 	xor	%r11,%r11
 
+#ifdef OPENSSL_PIC
+	mov     OPENSSL_ia32cap_P\@GOTPCREL(%rip),$idx#d
+#else
 	mov	OPENSSL_ia32cap_P(%rip),$idx#d
+#endif
 	bt	\$20,$idx#d	# RC4_CHAR?
 	jc	.Lc1stloop
 	jmp	.Lw1stloop
@@ -506,7 +514,11 @@
 .align	16
 RC4_options:
 	lea	.Lopts(%rip),%rax
+#ifdef OPENSSL_PIC
+	mov	OPENSSL_ia32cap_P\@GOTPCREL(%rip),%edx
+#else
 	mov	OPENSSL_ia32cap_P(%rip),%edx
+#endif
 	bt	\$20,%edx
 	jc	.L8xchar
 	bt	\$30,%edx
Index: openssl-1.0.1/crypto/perlasm/cbc.pl
===================================================================
--- openssl-1.0.1.orig/crypto/perlasm/cbc.pl	2011-07-13 06:22:46.000000000 +0000
+++ openssl-1.0.1/crypto/perlasm/cbc.pl	2012-03-17 09:20:04.000000000 +0000
@@ -122,7 +122,11 @@
 	&mov(&DWP($data_off,"esp","",0),	"eax");	# put in array for call
 	&mov(&DWP($data_off+4,"esp","",0),	"ebx");	#
 
-	&call($enc_func);
+	&call	(&label("pic_point0"));
+	&set_label("pic_point0");
+	&blindpop("ebx");
+	&add	("ebx", "\$_GLOBAL_OFFSET_TABLE_+[.-" . &label("pic_point0") . "]");
+	&call("$enc_func\@PLT");
 
 	&mov("eax",	&DWP($data_off,"esp","",0));
 	&mov("ebx",	&DWP($data_off+4,"esp","",0));
@@ -185,7 +189,11 @@
 	&mov(&DWP($data_off,"esp","",0),	"eax");	# put in array for call
 	&mov(&DWP($data_off+4,"esp","",0),	"ebx");	#
 
-	&call($enc_func);
+	&call	(&label("pic_point1"));
+	&set_label("pic_point1");
+	&blindpop("ebx");
+	&add	("ebx", "\$_GLOBAL_OFFSET_TABLE_+[.-" . &label("pic_point1") . "]");
+	&call("$enc_func\@PLT");
 
 	&mov("eax",	&DWP($data_off,"esp","",0));
 	&mov("ebx",	&DWP($data_off+4,"esp","",0));
@@ -218,7 +226,11 @@
 	&mov(&DWP($data_off,"esp","",0),	"eax");	# put back
 	&mov(&DWP($data_off+4,"esp","",0),	"ebx");	#
 
-	&call($dec_func);
+	&call	(&label("pic_point2"));
+	&set_label("pic_point2");
+	&blindpop("ebx");
+	&add	("ebx", "\$_GLOBAL_OFFSET_TABLE_+[.-" . &label("pic_point2") . "]");
+	&call("$dec_func\@PLT");
 
 	&mov("eax",	&DWP($data_off,"esp","",0));	# get return
 	&mov("ebx",	&DWP($data_off+4,"esp","",0));	#
@@ -261,7 +273,11 @@
 	&mov(&DWP($data_off,"esp","",0),	"eax");	# put back
 	&mov(&DWP($data_off+4,"esp","",0),	"ebx");	#
 
-	&call($dec_func);
+	&call	(&label("pic_point3"));
+	&set_label("pic_point3");
+	&blindpop("ebx");
+	&add	("ebx", "\$_GLOBAL_OFFSET_TABLE_+[.-" . &label("pic_point3") . "]");
+	&call("$dec_func\@PLT");
 
 	&mov("eax",	&DWP($data_off,"esp","",0));	# get return
 	&mov("ebx",	&DWP($data_off+4,"esp","",0));	#
Index: openssl-1.0.1/crypto/perlasm/x86_64-xlate.pl
===================================================================
--- openssl-1.0.1.orig/crypto/perlasm/x86_64-xlate.pl	2012-03-17 09:19:38.000000000 +0000
+++ openssl-1.0.1/crypto/perlasm/x86_64-xlate.pl	2012-03-17 09:20:04.000000000 +0000
@@ -786,7 +786,7 @@
 
     chomp($line);
 
-    $line =~ s|[#!].*$||;	# get rid of asm-style comments...
+#    $line =~ s|[#!].*$||;	# get rid of asm-style comments...
     $line =~ s|/\*.*\*/||;	# ... and C-style comments...
     $line =~ s|^\s+||;		# ... and skip white spaces in beginning
 
Index: openssl-1.0.1/crypto/perlasm/x86gas.pl
===================================================================
--- openssl-1.0.1.orig/crypto/perlasm/x86gas.pl	2011-12-09 19:16:35.000000000 +0000
+++ openssl-1.0.1/crypto/perlasm/x86gas.pl	2012-03-17 09:30:54.000000000 +0000
@@ -218,6 +218,15 @@
     elsif ($::elf)
     {	$initseg.=<<___;
 .section	.init
+#ifdef OPENSSL_PIC
+	pushl	%ebx
+	call	.pic_point0
+.pic_point0:
+	popl	%ebx
+	addl	\$_GLOBAL_OFFSET_TABLE_+[.-.pic_point0],%ebx
+	call	$f\@PLT
+	popl	%ebx
+#else
 	call	$f
 ___
     }
Index: openssl-1.0.1/crypto/aes/asm/aes-x86_64.pl
===================================================================
--- openssl-1.0.1.orig/crypto/aes/asm/aes-x86_64.pl	2011-11-14 21:01:17.000000000 +0000
+++ openssl-1.0.1/crypto/aes/asm/aes-x86_64.pl	2012-03-17 09:20:04.000000000 +0000
@@ -1678,7 +1678,11 @@
 	lea	.LAES_Td(%rip),$sbox
 .Lcbc_picked_te:
 
+#ifdef OPENSSL_PIC
+	mov	OPENSSL_ia32cap_P\@GOTPCREL(%rip),%r10d
+#else
 	mov	OPENSSL_ia32cap_P(%rip),%r10d
+#endif
 	cmp	\$$speed_limit,%rdx
 	jb	.Lcbc_slow_prologue
 	test	\$15,%rdx
Index: openssl-1.0.1/crypto/aes/Makefile
===================================================================
--- openssl-1.0.1.orig/crypto/aes/Makefile	2011-11-14 20:42:21.000000000 +0000
+++ openssl-1.0.1/crypto/aes/Makefile	2012-03-17 09:20:04.000000000 +0000
@@ -55,7 +55,7 @@
 aesni-x86.s:	asm/aesni-x86.pl ../perlasm/x86asm.pl
 	$(PERL) asm/aesni-x86.pl $(PERLASM_SCHEME) $(CFLAGS) $(PROCESSOR) > $@
 
-aes-x86_64.s: asm/aes-x86_64.pl
+aes-x86_64.S: asm/aes-x86_64.pl
 	$(PERL) asm/aes-x86_64.pl $(PERLASM_SCHEME) > $@
 vpaes-x86_64.s:	asm/vpaes-x86_64.pl
 	$(PERL) asm/vpaes-x86_64.pl $(PERLASM_SCHEME) > $@
