Index: crypto/aes/asm/vpaes-x86.pl
===================================================================
RCS file: /e/openssl/cvs/openssl/crypto/aes/asm/vpaes-x86.pl,v
retrieving revision 1.7
diff -u -w -r1.7 vpaes-x86.pl
--- a/crypto/aes/asm/vpaes-x86.pl	15 Dec 2011 22:20:04 -0000	1.7
+++ a/crypto/aes/asm/vpaes-x86.pl	31 Mar 2012 10:06:51 -0000
@@ -856,6 +856,7 @@
 	&sub	($round,16);
 	&mov	(&DWP(4,"esp"),$key)		# save key
 	&mov	(&DWP(8,"esp"),$const);		# save ivp
+	&jc	(&label("cbc_done"));
 	&mov	($out,$round);			# $out works as $len
 
 	&lea	($const,&DWP(&label("_vpaes_consts")."+0x30-".&label("pic_point")));
